File: structures/kmp.py
def prefix_function(s):
    q = 0
    n = len(s)
    
    pi = [0 for i in range(n)]
    
    for i in range(1, n):
        while q > 0 and s[i] != s[q]:
            q = pi[q - 1]
            
        if s[q] == s[i]:
            q += 1
            
        pi[i] = q
        
    return pi

def kmp(pat, txt):
    text = pat + '$' + txt
    pi = prefix_function(text)
    
    ans = 0
    n = len(pat)
    
    for i in range(n + 1, len(pi)):
        if pi[i] == n:
            ans += 1
            
    return ans

File: structures/kmp.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> prefix_function(const string &pat)
{
    int n = (int)pat.size();
    vector<int> pi(n);

    for (int i = 1; i < n; i++)
    {
        int j = pi[i - 1];

        while (j > 0 && pat[i] != pat[j])
            j = pi[j - 1];

        if (pat[i] == pat[j])
            j++;

        pi[i] = j;
    }
    return pi;
}

int kmp(string pat, string txt)
{
    int ans = 0;
    int n = pat.size();

    vector<int> pi = prefix_function(pat + '$' + txt);

    for (int i = n + 1; i < pi.size(); i++)
    {
        if (pi[i] == n)
            ans++;
    }

    return ans;
}